File: java/org.financer.client.javafx/src/main/java/org/financer/client/javafx/main/FinancerController.java
package org.financer.client.javafx.main;

import com.jfoenix.controls.JFXButton;
import com.jfoenix.controls.JFXHamburger;
import com.jfoenix.controls.JFXSnackbar;
import com.jfoenix.transitions.hamburger.HamburgerSlideCloseTransition;
import javafx.animation.ParallelTransition;
import javafx.animation.TranslateTransition;
import javafx.application.Platform;
import javafx.fxml.FXML;
import javafx.fxml.FXMLLoader;
import javafx.fxml.Initializable;
import javafx.geometry.Insets;
import javafx.scene.control.Button;
import javafx.scene.control.Label;
import javafx.scene.control.MenuButton;
import javafx.scene.control.MenuItem;
import javafx.scene.input.MouseEvent;
import javafx.scene.layout.BorderPane;
import javafx.scene.layout.StackPane;
import javafx.scene.layout.VBox;
import javafx.stage.Stage;
import javafx.util.Duration;
import org.controlsfx.glyphfont.FontAwesome;
import org.controlsfx.glyphfont.GlyphFont;
import org.controlsfx.glyphfont.GlyphFontRegistry;
import org.financer.client.connection.ServerRequest;
import org.financer.client.format.I18N;
import org.financer.client.javafx.dialogs.FinancerExceptionDialog;
import org.financer.client.javafx.local.LocalStorageImpl;
import org.financer.client.local.Application;

import java.io.IOException;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;
import java.util.ResourceBundle;
import java.util.concurrent.Executors;
import java.util.logging.Level;
import java.util.logging.Logger;

public class FinancerController implements Initializable, Application {

    @FXML
    public BorderPane rootLayout;
    @FXML
    public BorderPane header;
    private static VBox loadingBox;
    @FXML
    public Button overviewTabBtn;
    @FXML
    public JFXButton overviewTabIconBtn;
    @FXML
    public JFXButton transactionsTabBtn;
    @FXML
    public JFXButton transactionsTabIconBtn;
    @FXML
    public JFXButton statisticsTabBtn;
    @FXML
    public JFXButton statisticsTabIconBtn;
    @FXML
    public JFXButton profileTabBtn;
    @FXML
    public JFXButton profileTabIconBtn;
    @FXML
    public JFXButton settingTabBtn;
    @FXML
    public JFXButton settingTabIconBtn;
    @FXML
    public MenuButton accountMenuBtn;
    @FXML
    public MenuItem logoutBtn;
    @FXML
    public JFXHamburger hamburgerBtn;
    @FXML
    public Label contentLabel;
    @FXML
    public Label offlineLabel;
    @FXML
    public VBox navigationBox;
    @FXML
    public BorderPane contentPane;

    private static Application instance;

    private static Thread initializationThread = null;

    private static final Logger logger = Logger.getLogger("FinancerApplication");


    private ResourceBundle resourceBundle;
    private LocalStorageImpl localStorage = (LocalStorageImpl) LocalStorageImpl.getInstance();
    private JFXSnackbar snackbar;
    private boolean isNavigationBarHidden = false;

    @FXML
    public void initialize(URL location, ResourceBundle resources) {
        instance = this;
        ServerRequest.setApplication(this);
        ServerRequest.setLocalStorage(this.localStorage);

        try {
            ServerRequest.makeRequests(Executors.newCachedThreadPool());
        } catch (IOException e) {
            logger.log(Level.SEVERE, e.getMessage(), e);
        }

        User user = (User) this.localStorage.readObject("user");

        // setting up language
        this.resourceBundle = ResourceBundle.getBundle("Financer", user.getSettings().getLanguage());

        this.snackbar = new JFXSnackbar(this.rootLayout);

        try {
            loadingBox = FXMLLoader.load(getClass().getResource("/org/financer/client/javafx/main/views/loading.fxml"), this.resourceBundle);
        } catch (IOException e) {
            logger.log(Level.SEVERE, e.getMessage(), e);
        }

        this.accountMenuBtn.setText(user.getFullName());

        GlyphFont fontAwesome = GlyphFontRegistry.font("FontAwesome");
        this.accountMenuBtn.setGraphic(fontAwesome.create(FontAwesome.Glyph.USER));
        this.accountMenuBtn.setGraphicTextGap(10);
        this.logoutBtn.setGraphic(fontAwesome.create(FontAwesome.Glyph.SIGN_OUT));

        HamburgerSlideCloseTransition burgerTask = new HamburgerSlideCloseTransition(this.hamburgerBtn);
        burgerTask.setRate(1);
        Platform.runLater(() -> burgerTask.play());
        this.hamburgerBtn.setAnimation(burgerTask);
        this.hamburgerBtn.addEventHandler(MouseEvent.MOUSE_PRESSED, e -> {
            burgerTask.setRate(burgerTask.getRate() * -1);
            burgerTask.play();
        });

        handleShowOverviewContent();
    }

    public static Application getInstance() {
        return instance;
    }

    @Override
    public void showLoadingBox() {
        if (loadingBox != null) {
            loadingBox.setVisible(true);
        }
    }

    @Override
    public void hideLoadingBox() {
        if (loadingBox != null && loadingBox.isVisible()) {
            loadingBox.setVisible(false);
        }
    }

    @Override
    public void setOffline() {
//        Platform.runLater(() -> {
//            this.header.setStyle("-fx-background-color: #909ca8");
//            this.offlineLabel.setVisible(true);
//        });
    }

    @Override
    public void setOnline() {
//        Platform.runLater(() -> {
//            this.header.setStyle("-fx-background-color: #44a1a0");
//            this.offlineLabel.setVisible(false);
//        });
    }

    @Override
    public synchronized void showToast(MessageType messageType, String message) {
        Label messageLabel = new Label(message);
        messageLabel.getStyleClass().add(messageType.getName() + "-toast-label");
        this.snackbar.enqueue(new JFXSnackbar.SnackbarEvent(messageLabel));
    }

    @Override
    public void showErrorDialog(Exception exception) {
        Platform.runLater(() -> new FinancerExceptionDialog("Financer", exception));
    }

    public BorderPane getRootLayout() {
        return rootLayout;
    }

    public void handleShowOverviewContent() {
        this.loadFXML(getClass().getResource("/org/financer/client/javafx/main/views/overview.fxml"));
        this.removeSelectedStyleClass();
        this.overviewTabBtn.getStyleClass().add("selected");
        this.overviewTabIconBtn.getStyleClass().add("selected");
        this.contentLabel.setText(I18N.get("overview"));
    }


    public void handleShowTransactionsContent() {
        this.loadFXML(getClass().getResource("/org/financer/client/javafx/main/views/transactions.fxml"));
        this.removeSelectedStyleClass();
        this.transactionsTabBtn.getStyleClass().add("selected");
        this.transactionsTabIconBtn.getStyleClass().add("selected");
        this.contentLabel.setText(I18N.get("transactions"));
    }

    public void handleShowStatisticsContent() {
        this.loadFXML(getClass().getResource("/org/financer/client/javafx/main/views/statistics.fxml"));
        this.removeSelectedStyleClass();
        this.statisticsTabBtn.getStyleClass().add("selected");
        this.statisticsTabIconBtn.getStyleClass().add("selected");
        this.contentLabel.setText(I18N.get("statistics"));
    }

    public void handleShowProfileContent() {
        this.loadFXML(getClass().getResource("/org/financer/client/javafx/main/views/profile.fxml"));
        this.removeSelectedStyleClass();
        this.profileTabBtn.getStyleClass().add("selected");
        this.profileTabIconBtn.getStyleClass().add("selected");
        this.contentLabel.setText(I18N.get("profile"));
    }

    public void handleShowSettingsContent() {
        this.loadFXML(getClass().getResource("/org/financer/client/javafx/main/views/settings.fxml"));
        this.removeSelectedStyleClass();
        this.settingTabBtn.getStyleClass().add("selected");
        this.settingTabIconBtn.getStyleClass().add("selected");
        this.contentLabel.setText(I18N.get("settings"));
    }

    private void removeSelectedStyleClass() {
        final String navBtnClass = "nav-btn";
        this.overviewTabBtn.getStyleClass().clear();
        this.overviewTabBtn.getStyleClass().add(navBtnClass);
        this.transactionsTabBtn.getStyleClass().clear();
        this.transactionsTabBtn.getStyleClass().add(navBtnClass);
        this.statisticsTabBtn.getStyleClass().clear();
        this.statisticsTabBtn.getStyleClass().add(navBtnClass);
        this.profileTabBtn.getStyleClass().clear();
        this.profileTabBtn.getStyleClass().add(navBtnClass);
        this.settingTabBtn.getStyleClass().clear();
        this.settingTabBtn.getStyleClass().add(navBtnClass);

        this.overviewTabIconBtn.getStyleClass().clear();
        this.overviewTabIconBtn.getStyleClass().add(navBtnClass);
        this.transactionsTabIconBtn.getStyleClass().clear();
        this.transactionsTabIconBtn.getStyleClass().add(navBtnClass);
        this.statisticsTabIconBtn.getStyleClass().clear();
        this.statisticsTabIconBtn.getStyleClass().add(navBtnClass);
        this.profileTabIconBtn.getStyleClass().clear();
        this.profileTabIconBtn.getStyleClass().add(navBtnClass);
        this.settingTabIconBtn.getStyleClass().clear();
        this.settingTabIconBtn.getStyleClass().add(navBtnClass);
    }

    private void loadFXML(URL url) {
        try {
            StackPane stackPane = new StackPane();
            stackPane.getChildren().add(FXMLLoader.load(url, this.resourceBundle));
            stackPane.getChildren().add(loadingBox);
            this.contentPane.setCenter(stackPane);
            BorderPane.setMargin(stackPane, new Insets(20, 0, 0, 20));
        } catch (IOException e) {
            logger.log(Level.SEVERE, e.getMessage(), e);
        }
    }

    public void handleLogoutBtn() {
        handleLogout();
    }

    public static void handleLogout() {
        Stage stage = (Stage) loadingBox.getScene().getWindow();
        stage.close();

        LocalStorageImpl.getInstance().deleteAllData();

        try {
            new FinancerApplication().start(new Stage());
        } catch (IOException e) {
            logger.log(Level.SEVERE, e.getMessage(), e);
        }
    }

    public void onToggleNavigationBar() {
        this.overviewTabIconBtn.setVisible(!this.isNavigationBarHidden);
        this.transactionsTabIconBtn.setVisible(!this.isNavigationBarHidden);
        this.statisticsTabIconBtn.setVisible(!this.isNavigationBarHidden);
        this.profileTabIconBtn.setVisible(!this.isNavigationBarHidden);
        this.settingTabIconBtn.setVisible(!this.isNavigationBarHidden);

        Duration duration = new Duration(350);
        TranslateTransition t1 = new TranslateTransition(duration, this.navigationBox);
        TranslateTransition t2 = new TranslateTransition(duration, this.contentPane);

        List<TranslateTransition> tabIconBtnTransitions = new ArrayList<>();
        tabIconBtnTransitions.add(new TranslateTransition(duration, this.overviewTabIconBtn));
        tabIconBtnTransitions.add(new TranslateTransition(duration, this.transactionsTabIconBtn));
        tabIconBtnTransitions.add(new TranslateTransition(duration, this.statisticsTabIconBtn));
        tabIconBtnTransitions.add(new TranslateTransition(duration, this.profileTabIconBtn));
        tabIconBtnTransitions.add(new TranslateTransition(duration, this.settingTabIconBtn));
        List<TranslateTransition> tabBtnTransitions = new ArrayList<>();
        tabBtnTransitions.add(new TranslateTransition(duration, this.overviewTabBtn));
        tabBtnTransitions.add(new TranslateTransition(duration, this.transactionsTabBtn));
        tabBtnTransitions.add(new TranslateTransition(duration, this.statisticsTabBtn));
        tabBtnTransitions.add(new TranslateTransition(duration, this.profileTabBtn));
        tabBtnTransitions.add(new TranslateTransition(duration, this.settingTabBtn));

        if (!this.isNavigationBarHidden) {
            t1.setToX(-180);
            t2.setToX(-90);
            tabBtnTransitions.forEach(translateTransition -> translateTransition.setToX(-180));
            tabIconBtnTransitions.forEach(translateTransition -> {
                translateTransition.setFromX(-180);
                translateTransition.setToX(180);
            });
            this.overviewTabIconBtn.toFront();
            this.transactionsTabIconBtn.toFront();
            this.statisticsTabIconBtn.toFront();
            this.profileTabIconBtn.toFront();
            this.settingTabIconBtn.toFront();
        } else {
            t1.setToX(0);
            t2.setToX(0);
            tabBtnTransitions.forEach(translateTransition -> translateTransition.setToX(0));
            tabIconBtnTransitions.forEach(translateTransition -> translateTransition.setToX(0));
            this.overviewTabIconBtn.toBack();
            this.transactionsTabIconBtn.toBack();
            this.statisticsTabIconBtn.toBack();
            this.profileTabIconBtn.toBack();
            this.settingTabIconBtn.toBack();
        }
        ParallelTransition transition = new ParallelTransition(t1, t2);
        transition.getChildren().addAll(tabBtnTransitions);
        transition.getChildren().addAll(tabIconBtnTransitions);
        transition.play();

        this.isNavigationBarHidden = !this.isNavigationBarHidden;
    }

    public static Thread getInitializationThread() {
        return initializationThread;
    }

    public static void setInitializationThread(Thread initializationThread) {
        FinancerController.initializationThread = initializationThread;
    }
}

File: java/org.financer.client.javafx/src/main/kotlin/org/financer/client/javafx/main/FinancerController.kt
package org.financer.client.javafx.main

import com.jfoenix.controls.JFXButton
import com.jfoenix.controls.JFXHamburger
import com.jfoenix.controls.JFXSnackbar
import com.jfoenix.controls.JFXSnackbar.SnackbarEvent
import com.jfoenix.transitions.hamburger.HamburgerSlideCloseTransition
import javafx.animation.ParallelTransition
import javafx.animation.TranslateTransition
import javafx.application.Platform
import javafx.fxml.FXML
import javafx.fxml.FXMLLoader
import javafx.fxml.Initializable
import javafx.geometry.Insets
import javafx.scene.control.Button
import javafx.scene.control.Label
import javafx.scene.control.MenuButton
import javafx.scene.control.MenuItem
import javafx.scene.input.MouseEvent
import javafx.scene.layout.BorderPane
import javafx.scene.layout.StackPane
import javafx.scene.layout.VBox
import javafx.stage.Stage
import javafx.util.Duration
import org.controlsfx.glyphfont.FontAwesome
import org.controlsfx.glyphfont.GlyphFontRegistry
import org.financer.client.connection.ServerRequest
import org.financer.client.connection.ServerRequest.Companion.application
import org.financer.client.domain.model.user.User
import org.financer.client.format.FormatterImpl
import org.financer.client.format.I18N.get
import org.financer.client.javafx.dialogs.FinancerExceptionDialog
import org.financer.client.javafx.local.LocalStorageImpl
import org.financer.client.local.Application
import org.financer.shared.domain.model.value.objects.SettingPair
import java.io.IOException
import java.net.URL
import java.util.*
import java.util.function.Consumer
import java.util.logging.Level
import java.util.logging.Logger

class FinancerController : Initializable, Application {
    @FXML
    var rootLayout: BorderPane? = null

    @FXML
    var header: BorderPane? = null

    @FXML
    var overviewTabBtn: Button? = null

    @FXML
    var overviewTabIconBtn: JFXButton? = null

    @FXML
    var transactionsTabBtn: JFXButton? = null

    @FXML
    var transactionsTabIconBtn: JFXButton? = null

    @FXML
    var statisticsTabBtn: JFXButton? = null

    @FXML
    var statisticsTabIconBtn: JFXButton? = null

    @FXML
    var profileTabBtn: JFXButton? = null

    @FXML
    var profileTabIconBtn: JFXButton? = null

    @FXML
    var settingTabBtn: JFXButton? = null

    @FXML
    var settingTabIconBtn: JFXButton? = null

    @FXML
    var accountMenuBtn: MenuButton? = null

    @FXML
    var logoutBtn: MenuItem? = null

    @FXML
    var hamburgerBtn: JFXHamburger? = null

    @FXML
    var contentLabel: Label? = null

    @FXML
    var offlineLabel: Label? = null

    @FXML
    var navigationBox: VBox? = null

    @FXML
    var contentPane: BorderPane? = null
    private var resourceBundle: ResourceBundle? = null
    private val localStorage = LocalStorageImpl.getInstance() as LocalStorageImpl
    private var snackbar: JFXSnackbar? = null
    private var isNavigationBarHidden = false

    @FXML
    override fun initialize(location: URL, resources: ResourceBundle) {
        instance = this
        application = this
        ServerRequest.setLocalStorage(localStorage)
        val user = localStorage.readObject<User>("user")

        // setting up language
        resourceBundle = ResourceBundle.getBundle("Financer", user!!.getValueOrDefault<Any>(SettingPair.Property.LANGUAGE) as Locale)
        snackbar = JFXSnackbar(rootLayout)
        try {
            loadingBox = FXMLLoader.load(javaClass.getResource("/org/financer/client/javafx/main/views/loading.fxml"), resourceBundle)
        } catch (e: IOException) {
            logger.log(Level.SEVERE, e.message, e)
        }
        accountMenuBtn!!.text = FormatterImpl(localStorage).format(user.name!!)
        val fontAwesome = GlyphFontRegistry.font("FontAwesome")
        accountMenuBtn!!.graphic = fontAwesome.create(FontAwesome.Glyph.USER)
        accountMenuBtn!!.graphicTextGap = 10.0
        logoutBtn!!.graphic = fontAwesome.create(FontAwesome.Glyph.SIGN_OUT)
        val burgerTask = HamburgerSlideCloseTransition(hamburgerBtn)
        burgerTask.rate = 1.0
        Platform.runLater { burgerTask.play() }
        hamburgerBtn!!.animation = burgerTask
        hamburgerBtn!!.addEventHandler(MouseEvent.MOUSE_PRESSED) { e: MouseEvent? ->
            burgerTask.rate = burgerTask.rate * -1
            burgerTask.play()
        }
        handleShowOverviewContent()
    }

    override fun showLoadingBox() {
        if (loadingBox != null) {
            loadingBox!!.isVisible = true
        }
    }

    override fun hideLoadingBox() {
        if (loadingBox != null && loadingBox!!.isVisible) {
            loadingBox!!.isVisible = false
        }
    }

    override fun setOffline() {
//        Platform.runLater(() -> {
//            this.header.setStyle("-fx-background-color: #909ca8");
//            this.offlineLabel.setVisible(true);
//        });
    }

    override fun setOnline() {
//        Platform.runLater(() -> {
//            this.header.setStyle("-fx-background-color: #44a1a0");
//            this.offlineLabel.setVisible(false);
//        });
    }

    @Synchronized
    override fun showToast(messageType: Application.MessageType?, message: String?) {
        val messageLabel = Label(message)
        messageLabel.styleClass.add(messageType!!.type + "-toast-label")
        snackbar!!.enqueue(SnackbarEvent(messageLabel))
    }

    override fun showErrorDialog(exception: Exception?) {
        Platform.runLater { FinancerExceptionDialog("Financer", exception) }
    }

    fun handleShowOverviewContent() {
        loadFXML(javaClass.getResource("/org/financer/client/javafx/main/views/overview.fxml"))
        removeSelectedStyleClass()
        overviewTabBtn!!.styleClass.add("selected")
        overviewTabIconBtn!!.styleClass.add("selected")
        contentLabel!!.text = get("overview")
    }

    fun handleShowTransactionsContent() {
        loadFXML(javaClass.getResource("/org/financer/client/javafx/main/views/transactions.fxml"))
        removeSelectedStyleClass()
        transactionsTabBtn!!.styleClass.add("selected")
        transactionsTabIconBtn!!.styleClass.add("selected")
        contentLabel!!.text = get("transactions")
    }

    fun handleShowStatisticsContent() {
        loadFXML(javaClass.getResource("/org/financer/client/javafx/main/views/statistics.fxml"))
        removeSelectedStyleClass()
        statisticsTabBtn!!.styleClass.add("selected")
        statisticsTabIconBtn!!.styleClass.add("selected")
        contentLabel!!.text = get("statistics")
    }

    fun handleShowProfileContent() {
        loadFXML(javaClass.getResource("/org/financer/client/javafx/main/views/profile.fxml"))
        removeSelectedStyleClass()
        profileTabBtn!!.styleClass.add("selected")
        profileTabIconBtn!!.styleClass.add("selected")
        contentLabel!!.text = get("profile")
    }

    fun handleShowSettingsContent() {
        loadFXML(javaClass.getResource("/org/financer/client/javafx/main/views/settings.fxml"))
        removeSelectedStyleClass()
        settingTabBtn!!.styleClass.add("selected")
        settingTabIconBtn!!.styleClass.add("selected")
        contentLabel!!.text = get("settings")
    }

    private fun removeSelectedStyleClass() {
        val navBtnClass = "nav-btn"
        overviewTabBtn!!.styleClass.clear()
        overviewTabBtn!!.styleClass.add(navBtnClass)
        transactionsTabBtn!!.styleClass.clear()
        transactionsTabBtn!!.styleClass.add(navBtnClass)
        statisticsTabBtn!!.styleClass.clear()
        statisticsTabBtn!!.styleClass.add(navBtnClass)
        profileTabBtn!!.styleClass.clear()
        profileTabBtn!!.styleClass.add(navBtnClass)
        settingTabBtn!!.styleClass.clear()
        settingTabBtn!!.styleClass.add(navBtnClass)
        overviewTabIconBtn!!.styleClass.clear()
        overviewTabIconBtn!!.styleClass.add(navBtnClass)
        transactionsTabIconBtn!!.styleClass.clear()
        transactionsTabIconBtn!!.styleClass.add(navBtnClass)
        statisticsTabIconBtn!!.styleClass.clear()
        statisticsTabIconBtn!!.styleClass.add(navBtnClass)
        profileTabIconBtn!!.styleClass.clear()
        profileTabIconBtn!!.styleClass.add(navBtnClass)
        settingTabIconBtn!!.styleClass.clear()
        settingTabIconBtn!!.styleClass.add(navBtnClass)
    }

    private fun loadFXML(url: URL) {
        try {
            val stackPane = StackPane()
            stackPane.children.add(FXMLLoader.load(url, resourceBundle))
            stackPane.children.add(loadingBox)
            contentPane!!.center = stackPane
            BorderPane.setMargin(stackPane, Insets(20.0, 0.0, 0.0, 20.0))
        } catch (e: IOException) {
            logger.log(Level.SEVERE, e.message, e)
        }
    }

    fun handleLogoutBtn() {
        handleLogout()
    }

    fun onToggleNavigationBar() {
        overviewTabIconBtn!!.isVisible = !isNavigationBarHidden
        transactionsTabIconBtn!!.isVisible = !isNavigationBarHidden
        statisticsTabIconBtn!!.isVisible = !isNavigationBarHidden
        profileTabIconBtn!!.isVisible = !isNavigationBarHidden
        settingTabIconBtn!!.isVisible = !isNavigationBarHidden
        val duration = Duration(350.0)
        val t1 = TranslateTransition(duration, navigationBox)
        val t2 = TranslateTransition(duration, contentPane)
        val tabIconBtnTransitions: MutableList<TranslateTransition> = ArrayList()
        tabIconBtnTransitions.add(TranslateTransition(duration, overviewTabIconBtn))
        tabIconBtnTransitions.add(TranslateTransition(duration, transactionsTabIconBtn))
        tabIconBtnTransitions.add(TranslateTransition(duration, statisticsTabIconBtn))
        tabIconBtnTransitions.add(TranslateTransition(duration, profileTabIconBtn))
        tabIconBtnTransitions.add(TranslateTransition(duration, settingTabIconBtn))
        val tabBtnTransitions: MutableList<TranslateTransition> = ArrayList()
        tabBtnTransitions.add(TranslateTransition(duration, overviewTabBtn))
        tabBtnTransitions.add(TranslateTransition(duration, transactionsTabBtn))
        tabBtnTransitions.add(TranslateTransition(duration, statisticsTabBtn))
        tabBtnTransitions.add(TranslateTransition(duration, profileTabBtn))
        tabBtnTransitions.add(TranslateTransition(duration, settingTabBtn))
        if (!isNavigationBarHidden) {
            t1.toX = -180.0
            t2.toX = -90.0
            tabBtnTransitions.forEach(Consumer { translateTransition: TranslateTransition -> translateTransition.toX = -180.0 })
            tabIconBtnTransitions.forEach(Consumer { translateTransition: TranslateTransition ->
                translateTransition.fromX = -180.0
                translateTransition.toX = 180.0
            })
            overviewTabIconBtn!!.toFront()
            transactionsTabIconBtn!!.toFront()
            statisticsTabIconBtn!!.toFront()
            profileTabIconBtn!!.toFront()
            settingTabIconBtn!!.toFront()
        } else {
            t1.toX = 0.0
            t2.toX = 0.0
            tabBtnTransitions.forEach(Consumer { translateTransition: TranslateTransition -> translateTransition.toX = 0.0 })
            tabIconBtnTransitions.forEach(Consumer { translateTransition: TranslateTransition -> translateTransition.toX = 0.0 })
            overviewTabIconBtn!!.toBack()
            transactionsTabIconBtn!!.toBack()
            statisticsTabIconBtn!!.toBack()
            profileTabIconBtn!!.toBack()
            settingTabIconBtn!!.toBack()
        }
        val transition = ParallelTransition(t1, t2)
        transition.children.addAll(tabBtnTransitions)
        transition.children.addAll(tabIconBtnTransitions)
        transition.play()
        isNavigationBarHidden = !isNavigationBarHidden
    }

    companion object {
        private var loadingBox: VBox? = null
        var instance: Application? = null
            private set
        private val logger = Logger.getLogger("FinancerApplication")

        fun handleLogout() {
            val stage = loadingBox!!.scene.window as Stage
            stage.close()
            LocalStorageImpl.getInstance().deleteAllData()
            try {
                FinancerApplication().start(Stage())
            } catch (e: IOException) {
                logger.log(Level.SEVERE, e.message, e)
            }
        }
    }
}